// devexpress-reporting/designer/controls/metadata/properties/editOptions.js
/**
* DevExpress HTML/JS Reporting (designer\controls\metadata\properties\editOptions.js)
* Version:  23.2.3
* Build date: Dec 1, 2023
* Copyright (c) 2012 - 2023 Developer Express Inc. ALL RIGHTS RESERVED
* License: https://www.devexpress.com/Support/EULAs/universal.xml
*/
import { extend } from '@devexpress/analytics-core/analytics-internal';
import { editorTemplates } from '@devexpress/analytics-core/analytics-widgets';
import { EditOptions, TextEditOptions } from '../../properties/editOptions';
export var editOptions = {
    propertyName: 'editOptions',
    modelName: 'EditOptions',
    displayName: 'Edit Options', localizationId: 'DevExpress.XtraReports.UI.XRLabel.EditOptions',
    editor: editorTemplates.getEditor('objecteditor'),
    from: (model, serializer) => new EditOptions(model, serializer),
    toJsonObject: (value, serializer) => serializer.serialize(value)
};
export var textEditOptions = extend({}, editOptions, {
    propertyName: 'textEditOptions',
    from: (model, serializer) => new TextEditOptions(model, serializer)
});
