// devexpress-reporting/designer/controls/metadata/properties/metadataGroups.js
/**
* DevExpress HTML/JS Reporting (designer\controls\metadata\properties\metadataGroups.js)
* Version:  23.2.3
* Build date: Dec 1, 2023
* Copyright (c) 2012 - 2023 Developer Express Inc. ALL RIGHTS RESERVED
* License: https://www.devexpress.com/Support/EULAs/universal.xml
*/
import { expressionBindings } from './expressionBinding';
import { formattingRuleLinks } from './formattingRulesLink';
import { accessibleDescription, angle, backColor, bookmark, bookmarkParent, borderColor, borderDashStyle, borders, borderWidth, canGrow, canPublish, canShrink, dataAdapter, dataMember, dataSource, dpi, expressionableFont, filterString, filterStringEditable, foreColor, keepTogetherDefaultValueFalse, location, lockedInUserDesigner, multiline, name, navigateUrl, nullValueText, paddingGroup, processDuplicatesMode, processDuplicatesTarget, processNullValues, reportPrintOptions, rtl, size, summary, tag, target, text, textAlignment, textArea, textFitMode, textFormatString, visible, wordWrap, xlsxFormatString } from './metadata';
import { interactiveSorting } from './sortingOptions';
import { evenStyleName, oddStyleName, styleName, stylePriority, stylesObj } from './style';
export var sizeLocation = [size, location];
export var bordersProperties = [borders, borderWidth, borderDashStyle, borderColor];
export var baseControlProperties = [name, visible, dpi, lockedInUserDesigner, tag, expressionBindings];
export var commonBandProperties = [backColor, formattingRuleLinks].concat(baseControlProperties, bordersProperties, paddingGroup);
export var commonControlProperties = [styleName, evenStyleName, accessibleDescription, oddStyleName, stylesObj, stylePriority, canPublish].concat(commonBandProperties);
export var fontGroup = [expressionableFont, foreColor];
export var bookmarkGroup = [bookmark, bookmarkParent];
export var navigationGroup = [navigateUrl, target].concat(bookmarkGroup);
export var datasourcePrintOptionsGroup = [dataSource, dataMember, dataAdapter, filterString, filterStringEditable, reportPrintOptions];
export var processGroup = [processDuplicatesMode, processDuplicatesTarget, processNullValues];
export var canGrowShrinkGroup = [canGrow, canShrink];
export var labelGroup = [textAlignment, text, textArea, textFormatString, textFitMode, nullValueText,
    keepTogetherDefaultValueFalse, summary, multiline, angle, wordWrap, xlsxFormatString, rtl, interactiveSorting
].concat(commonControlProperties, fontGroup, navigationGroup, canGrowShrinkGroup, processGroup);
export var unknownSerializationsInfo = [].concat(baseControlProperties, sizeLocation);
