// devexpress-reporting/viewer/settings.js
/**
* DevExpress HTML/JS Reporting (viewer\settings.js)
* Version:  23.2.3
* Build date: Dec 1, 2023
* Copyright (c) 2012 - 2023 Developer Express Inc. ALL RIGHTS RESERVED
* License: https://www.devexpress.com/Support/EULAs/universal.xml
*/
import { createGlobalModuleVariableFunc, NotifyAboutWarning, NotifyType, ShowMessage } from '@devexpress/analytics-core/analytics-internal-native';
export var EditablePreviewEnabled = createGlobalModuleVariableFunc(true);
export var SearchAvailable = createGlobalModuleVariableFunc(true);
export var ReportServerInvokeUri = '/RSWebDocumentViewerApi/Invoke';
export var ReportServerExportUri = '/RSWebDocumentViewerApi/Download';
export var AsyncExportApproach = createGlobalModuleVariableFunc(false);
export var MessageHandler = createGlobalModuleVariableFunc({
    processError: (message, showForUser, prefix = '', element) => {
        showForUser && ShowMessage(message.substr(prefix.length), undefined, undefined, undefined, undefined, element);
        NotifyAboutWarning(message, false);
    },
    processMessage: (message, showForUser, element) => { showForUser && ShowMessage(message, NotifyType.success, 10000, undefined, undefined, element); },
    processWarning: (message, showForUser, element) => { showForUser && ShowMessage(message, undefined, undefined, undefined, undefined, element); }
});
export var HandlerUri = createGlobalModuleVariableFunc('DXXRDV.axd');
export var previewDefaultResolution = createGlobalModuleVariableFunc(96);
export var ReportServerDownloadUri = createGlobalModuleVariableFunc('');
export var PollingDelay = createGlobalModuleVariableFunc(300);
export var TimeOut = createGlobalModuleVariableFunc(105000);
export var PreloadedPagesOffset = createGlobalModuleVariableFunc(15);
export var MultipageScrollingThrottle = createGlobalModuleVariableFunc(150);
