// devexpress-reporting/designer/dataObjects/metadata/parameters/rangeSettings.js
/**
* DevExpress HTML/JS Reporting (designer\dataObjects\metadata\parameters\rangeSettings.js)
* Version:  23.2.3
* Build date: Dec 1, 2023
* Copyright (c) 2012 - 2023 Developer Express Inc. ALL RIGHTS RESERVED
* License: https://www.devexpress.com/Support/EULAs/universal.xml
*/
import { editorTemplates, PropertyGridEditorFlat } from '@devexpress/analytics-core/analytics-widgets';
import { parameterExpressionSerializationInfo, parameterNameSerializationInfo, parameterValueSerializationInfo } from './parameter';
import { parameterExpressionBindings } from './parameterExpressionBinding';
export var rangeEditor = {
    custom: 'dxrd-parameters-range-grid', editorType: PropertyGridEditorFlat
};
export var rangeBoundaryParameterInfos = [
    parameterNameSerializationInfo, parameterValueSerializationInfo,
    parameterExpressionBindings, parameterExpressionSerializationInfo,
];
var startParameter = {
    propertyName: 'startParameter', modelName: '@StartParameter', displayName: 'Start Parameter', localizationId: 'DevExpress.XtraReports.Parameters.RangeSettings.StartParameter',
    editor: editorTemplates.getEditor('objecteditor'), link: true
};
var endParameter = {
    propertyName: 'endParameter', modelName: '@EndParameter', displayName: 'End Parameter', localizationId: 'DevExpress.XtraReports.Parameters.RangeSettings.EndParameter',
    editor: editorTemplates.getEditor('objecteditor'), link: true
};
export var rangeSettingsInfos = [startParameter, endParameter];
