// devexpress-reporting/common/metadata.js
/**
* DevExpress HTML/JS Reporting (common\metadata.js)
* Version:  23.2.3
* Build date: Dec 1, 2023
* Copyright (c) 2012 - 2023 Developer Express Inc. ALL RIGHTS RESERVED
* License: https://www.devexpress.com/Support/EULAs/universal.xml
*/
import { PaddingModel } from '@devexpress/analytics-core/analytics-elements-native';
import { colorFromString, colorToString, floatFromModel } from '@devexpress/analytics-core/analytics-utils-native';
export var previewBackColor = { propertyName: 'backColor', modelName: '@BackColor', from: colorFromString, toJsonObject: colorToString };
export var previewSides = { propertyName: 'borders', modelName: '@Sides' };
export var previewBorderColor = { propertyName: 'borderColor', modelName: '@BorderColor', from: colorFromString, toJsonObject: colorToString };
export var previewBorderStyle = { propertyName: 'borderStyle', modelName: '@BorderStyle' };
export var previewBorderDashStyle = { propertyName: 'borderDashStyle', modelName: '@BorderDashStyle' };
export var previewBorderWidth = { propertyName: 'borderWidth', modelName: '@BorderWidthSerializable', from: floatFromModel };
export var previewForeColor = { propertyName: 'foreColor', modelName: '@ForeColor', from: colorFromString, toJsonObject: colorToString };
export var previewFont = { propertyName: 'font', modelName: '@Font' };
export var previewPadding = { propertyName: 'padding', modelName: '@Padding', from: PaddingModel.from };
export var previewTextAlignment = { propertyName: 'textAlignment', modelName: '@TextAlignment' };
export var brickStyleSerializationsInfo = [
    previewBackColor,
    previewSides,
    previewBorderColor,
    previewBorderStyle,
    previewBorderDashStyle,
    previewBorderWidth,
    previewForeColor,
    previewFont,
    previewPadding,
    previewTextAlignment
];
