// devexpress-reporting/designer/controls/metadata/crosstab/defenitions.js
/**
* DevExpress HTML/JS Reporting (designer\controls\metadata\crosstab\defenitions.js)
* Version:  23.2.3
* Build date: Dec 1, 2023
* Copyright (c) 2012 - 2023 Developer Express Inc. ALL RIGHTS RESERVED
* License: https://www.devexpress.com/Support/EULAs/universal.xml
*/
import { floatFromModel, parseBool } from '@devexpress/analytics-core/analytics-utils';
import { autoSizeMode } from './xrCrossTabCell';
export var crossTabCellWidth = { propertyName: 'width', modelName: '@Width', localizationId: 'AnalyticsCoreStringId.SizeF.Width', defaultVal: 100, from: floatFromModel };
var autoWidthMode = { propertyName: 'autoWidthMode', modelName: '@AutoWidthMode', defaultVal: 'None', valuesArray: autoSizeMode };
var visible = { propertyName: 'visible', modelName: '@Visible', defaultVal: true, from: parseBool };
export var crossTabColumnDefinitionInfo = [crossTabCellWidth, autoWidthMode, visible];
export var crossTabCellHeight = { propertyName: 'height', modelName: '@Height', displayName: 'Height', localizationId: 'AnalyticsCoreStringId.SizeF.Height', defaultVal: 25, from: floatFromModel };
var autoHeightMode = { propertyName: 'autoHeightMode', modelName: '@AutoHeightMode', defaultVal: 'None', valuesArray: autoSizeMode };
export var crossTabRowDefinitionInfo = [crossTabCellHeight, autoHeightMode, visible];
export var rowDefinitions = { propertyName: '_rowDefinitions', modelName: 'RowDefinitions', array: true, alwaysSerialize: true };
export var columnDefinitions = { propertyName: '_columnDefinitions', modelName: 'ColumnDefinitions', array: true, alwaysSerialize: true };
