// devexpress-reporting/designer/controls/metadata/xrTableOfContents.js
/**
* DevExpress HTML/JS Reporting (designer\controls\metadata\xrTableOfContents.js)
* Version:  23.2.3
* Build date: Dec 1, 2023
* Copyright (c) 2012 - 2023 Developer Express Inc. ALL RIGHTS RESERVED
* License: https://www.devexpress.com/Support/EULAs/universal.xml
*/
import { Size } from '@devexpress/analytics-core/analytics-elements';
import { editorTemplates } from '@devexpress/analytics-core/analytics-widgets';
import * as $ from 'jquery';
import { designerEditorTemplates } from '../../widgets/editorTemplates';
import { accessibleDescription, canPublish, location, rtl, text, textAlignment } from './properties/metadata';
import { commonControlProperties } from './properties/metadataGroups';
import { baseTocLevelSerializationsInfo, tocLevelSerializationsInfo } from './xrTableOfContentsLevel';
var size = { propertyName: 'size', modelName: '@SizeF', from: Size.fromString };
var formattingRuleLinks = {
    propertyName: 'formattingRuleLinks', modelName: 'FormattingRuleLinks'
};
export var tocTitleSerializationsInfo = [text, $.extend({}, textAlignment, { defaultVal: 'TopLeft' })].concat(baseTocLevelSerializationsInfo);
export var tocTitle = { propertyName: 'levelTitle', modelName: 'LevelTitle', displayName: 'Level Title', localizationId: 'DevExpress.XtraReports.UI.XRTableOfContents.LevelTitle', info: tocTitleSerializationsInfo, editor: editorTemplates.getEditor('objecteditor') };
export var tocLevelDefault = { propertyName: 'levelDefault', modelName: 'LevelDefault', displayName: 'Level Default', localizationId: 'DevExpress.XtraReports.UI.XRTableOfContents.LevelDefault', info: tocLevelSerializationsInfo, editor: editorTemplates.getEditor('objecteditor') };
export var maxNestingLevel = { propertyName: 'maxNestingLevel', modelName: '@MaxNestingLevel', defaultVal: 0, displayName: 'Max Nesting Level', localizationId: 'DevExpress.XtraReports.UI.XRTableOfContents.MaxNestingLevel', editor: editorTemplates.getEditor('numeric'), editorOptions: { min: 0 } };
export var tocLevels = {
    propertyName: 'levels',
    modelName: 'Levels',
    displayName: 'Levels', localizationId: 'DevExpress.XtraReports.UI.XRTableOfContents.Levels',
    array: true,
    editor: designerEditorTemplates.getEditor('toclevel'),
    template: '#dxrd-collectionItemWithAccordion'
};
var tocProperties = commonControlProperties.filter((item) => { return item !== canPublish && item != accessibleDescription; });
export var tocSerializationsInfo = [formattingRuleLinks, size, location, tocTitle, tocLevels, tocLevelDefault, maxNestingLevel, rtl].concat(tocProperties);
