// devexpress-reporting/viewer/exportOptions/exportOptionsModel.js
/**
* DevExpress HTML/JS Reporting (viewer\exportOptions\exportOptionsModel.js)
* Version:  23.2.3
* Build date: Dec 1, 2023
* Copyright (c) 2012 - 2023 Developer Express Inc. ALL RIGHTS RESERVED
* License: https://www.devexpress.com/Support/EULAs/universal.xml
*/
import { $unwrap, PropertyGridKeyboardHelper } from '@devexpress/analytics-core/analytics-internal-native';
import { Disposable, TabInfoWithPropertyGrid, } from '@devexpress/analytics-core/analytics-utils-native';
import { ActionId, ExportFormatID } from '../constants';
export class ExportOptionsModel extends Disposable {
    constructor(reportPreview, enableKeyboardSupport) {
        super();
        this.actions = [];
        this._reportPreview = reportPreview;
        let exportDisabledChanged = null;
        let exportOptionsModelChanged = null;
        const getTabPanelVisible = () => !!reportPreview.exportOptionsModel && reportPreview.exportOptionsTabVisible !== false;
        const keyboardHelperViewModelAccessor = () => this.tabInfo.getViewModel().propertyGrid.editors;
        const keyboardHelper = enableKeyboardSupport ? new PropertyGridKeyboardHelper(keyboardHelperViewModelAccessor) : undefined;
        this.tabInfo = new TabInfoWithPropertyGrid({
            text: 'Export Options',
            template: 'dxrd-preview-export-options',
            propertyGridModel: reportPreview.exportOptionsModel,
            model: undefined,
            keyboardHelper,
            localizationId: 'DevExpress.XtraPrinting.ExportOptions',
            imageClassName: 'properties',
            imageTemplateName: 'dxrd-svg-tabs-properties',
            visible: getTabPanelVisible(),
            engineType: 'native'
        });
        this.addDisposable(reportPreview.events.on('exportOptionsModelChanged', (args) => {
            if (reportPreview.exportOptionsModel)
                this.tabInfo.propertyGridModel = reportPreview.exportOptionsModel;
            exportOptionsModelChanged && exportOptionsModelChanged();
            this.tabInfo.visible = getTabPanelVisible();
        }), reportPreview.events.on('exportOptionsTabVisibleChanged', (args) => {
            this.tabInfo.visible = getTabPanelVisible();
        }), reportPreview.events.on('exportDisabledChanged', (args) => {
            exportDisabledChanged && exportDisabledChanged(args.newValue);
        }));
        const getExportItems = () => {
            var result = this._getExportFormatItems();
            return [{
                    text: 'Export To',
                    textId: 'ASPxReportsStringId.WebDocumentViewer_ExportToText',
                    imageClassName: 'dxrd-image-export-to',
                    items: result,
                    get imageTemplateName() {
                        return actionExportTo.imageTemplateName;
                    },
                    set imageTemplateName(newVal) {
                        actionExportTo.imageTemplateName = newVal;
                    }
                }];
        };
        var actionExportTo = {
            id: ActionId.ExportTo,
            text: 'Export To',
            textId: 'ASPxReportsStringId.WebDocumentViewer_ExportToText',
            getDisabled: () => reportPreview.exportDisabled,
            setDisabledChangedEvent: (callback) => {
                exportDisabledChanged = callback;
                return () => exportDisabledChanged = undefined;
            },
            visible: true,
            items: getExportItems(),
            setItemsChangedEvent: (callback) => {
                exportOptionsModelChanged = () => callback(getExportItems());
                return () => exportOptionsModelChanged = undefined;
            },
            clickAction: (model) => {
                if (reportPreview.exportDisabled)
                    return;
                this._exportDocumentByFormat(model.itemData.format);
            },
            templateName: 'dxrd-preview-export-to',
            imageTemplateName: 'dxrd-svg-preview-export-export-to',
            eventHandlers: new ExportOptionsEventHandlers()
        };
        this.actions.push(actionExportTo);
        this._disposables.push(this.tabInfo);
    }
    _getExportFormatItems() {
        var result = [];
        var exportOptionsModel = this._reportPreview.exportOptionsModel;
        if (exportOptionsModel) {
            exportOptionsModel.pdf && result.push(ExportFormatID.PDF);
            exportOptionsModel.xls && result.push(ExportFormatID.XLS);
            exportOptionsModel.xlsx && result.push(ExportFormatID.XLSX);
            exportOptionsModel.rtf && result.push(ExportFormatID.RTF);
            exportOptionsModel.docx && result.push(ExportFormatID.DOCX);
            exportOptionsModel.mht && result.push(ExportFormatID.MHT);
            exportOptionsModel.html && result.push(ExportFormatID.HTML);
            exportOptionsModel.textExportOptions && result.push(ExportFormatID.Text);
            exportOptionsModel.csv && result.push(ExportFormatID.CSV);
            exportOptionsModel.image && result.push(ExportFormatID.Image);
        }
        return result;
    }
    _exportDocumentByFormat(format) {
        format && this._reportPreview.exportDocumentTo(format);
    }
    getActions(context) {
        return this.actions;
    }
    dispose() {
        super.dispose();
        this.disposeArray(this.actions);
        this.removeProperties();
    }
}
export class ExportOptionsEventHandlers {
    onSubmenuShowing(popupContainer, element) {
        return (e) => {
            e.submenu._overlay.option('container', popupContainer);
            e.submenu._overlay.option('focusStateEnabled', false);
            e.submenu._overlay.option('position', { my: 'left top', at: 'left bottom', of: element, collision: 'none', boundary: popupContainer });
            this._menuButton = e.component.option('focusedElement');
            e.submenu._attachHoverEvents();
        };
    }
    onSubmenuShown(e) {
        var submenu = e.submenu;
        submenu.registerKeyHandler('escape', (e) => submenu.hide());
        submenu.registerKeyHandler('leftArrow', (e) => submenu.hide());
        submenu.registerKeyHandler('rightArrow', (e) => submenu.hide());
    }
    onSubmenuHiding(e) {
        if (this._menuButton) {
            e.component.option('focusedElement', this._menuButton);
            e.component.focus();
        }
    }
    onItemRendered(e) {
        var _a;
        $unwrap(e.itemElement).setAttribute('aria-label', (_a = e.itemData) === null || _a === void 0 ? void 0 : _a.text);
    }
}
