// devexpress-reporting/viewer/internal/_actions.js
/**
* DevExpress HTML/JS Reporting (viewer\internal\_actions.js)
* Version:  23.2.3
* Build date: Dec 1, 2023
* Copyright (c) 2012 - 2023 Developer Express Inc. ALL RIGHTS RESERVED
* License: https://www.devexpress.com/Support/EULAs/universal.xml
*/
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
import { ActionId, ZoomAutoBy } from '../constants';
import { EditablePreviewEnabled } from '../settings';
import { getLocalization, Disposable } from '@devexpress/analytics-core/analytics-utils-native';
import { parseZoom, getParentContainer, ActionListsBase } from '@devexpress/analytics-core/analytics-internal-native';
import { createViewModelGenerator, mutableArray } from '@devexpress/analytics-core/analytics-serializer-native';
export class PreviewDesignerActions extends Disposable {
    constructor(reportPreview, fullscreen) {
        super();
        this.actions = [];
        var designAction = {
            id: ActionId.Design,
            text: 'Design',
            displayText: () => getLocalization('Design', 'ASPxReportsStringId.ToolBarItemText_Design'),
            imageClassName: 'dxrd-image-design',
            imageTemplateName: 'dxrd-svg-preview-report_designer',
            templateName: reportPreview.canSwitchToDesigner ? 'dxrd-toolbar-two-way-switch' : undefined,
            visible: reportPreview.canSwitchToDesigner,
            hotKey: { ctrlKey: true, keyCode: 68 },
            clickAction: () => {
                reportPreview.previewVisible = false;
                reportPreview.deactivate();
            }
        };
        designAction.contentData = {
            items: [
                { itemData: designAction, active: false },
                { getDisplayText: () => getLocalization('Preview', 'ASPxReportsStringId.ToolBarItemText_Preview'), active: true }
            ]
        };
        this.actions.push(designAction);
        const getFullScreenImageTemplateName = (fullscreen) => fullscreen ? 'dxrd-svg-toolbar-fullscreen-exit' : 'dxrd-svg-toolbar-fullscreen';
        this.actions.push({
            id: ActionId.FullScreen,
            text: getLocalization('Full Screen', 'ASPxReportsStringId.ToolBarItemText_FullScreen'),
            imageClassName: 'dxrd-image-fullscreen',
            imageTemplateName: getFullScreenImageTemplateName(fullscreen()),
            setImageTemplateNameChangedEvent: (callback) => {
                const handler = (newValue) => callback(getFullScreenImageTemplateName(newValue));
                const dispose = fullscreen.subscribe(handler);
                return () => dispose();
            },
            visible: !reportPreview.canSwitchToDesigner,
            selected: fullscreen(),
            setSelectedChangedEvent: (callback) => {
                const dispose = fullscreen.subscribe(callback);
                return () => dispose();
            },
            clickAction: () => {
                fullscreen(!fullscreen());
            }
        });
    }
    dispose() {
        super.dispose();
        this.removeProperties();
    }
    getActions(context) {
        return this.actions;
    }
}
export class ActionLists extends ActionListsBase {
    constructor(reportPreview, globalActionProviders, customizeActions, enabled) {
        super(enabled);
        this.updateToolbarItems = () => {
            var globalActions = [];
            this.globalActionProviders.forEach((actionProvider) => {
                globalActions.push.apply(globalActions, actionProvider.getActions(reportPreview));
            });
            customizeActions && customizeActions(globalActions);
            this.toolbarItems = globalActions;
        };
        this.globalActionProviders = globalActionProviders;
    }
    _isPaginationAction(action) {
        return 'pageItems' in action;
    }
    _isExportAction(action) {
        return 'eventHandlers' in action;
    }
    createActionViewModel(action, index) {
        const viewModel = super.createActionViewModel(action, index);
        if ('zoom' in action) {
            const selectBoxViewModel = createViewModelGenerator(viewModel)
                .configureProperty('widget', (property) => {
                property.inputAttr = {
                    'aria-label': getLocalization('Zoom', 'ASPxReportsStringId.WebDocumentViewer_AriaLabelZoomCombobox')
                };
            })
                .getViewModel();
            return selectBoxViewModel;
        }
        if (this._isPaginationAction(action)) {
            const selectBoxViewModel = createViewModelGenerator(viewModel)
                .generateProperty('templateName', action.templateName)
                .generateProperty('getPopupContainer', getParentContainer)
                .generateProperty('widget', createViewModelGenerator()
                .generateProperty('dataSource', action.pageItems)
                .generateProperty('displayExpr', action.displayExpr)
                .generateProperty('value', action.selectedItem)
                .generateProperty('searchEnabled', true)
                .generateProperty('opened', false)
                .generateProperty('onValueChanged', (e) => {
                action.selectItem(e.value);
            })
                .generateProperty('onKeyUp', (e) => {
                if (e.event.which !== 13)
                    return;
                if (action._isPageChanged(e.component.option('text'))) {
                    selectBoxViewModel.widget.opened = false;
                }
            })
                .generateProperty('searchMode', action.searchMode)
                .generateProperty('searchTimeout', action.searchTimeout)
                .generateProperty('itemTemplate', action.itemTemplate)
                .generateProperty('dropDownOptions', {
                wrapperAttr: {
                    'aria-label': getLocalization('Pager', 'ASPxReportsStringId.WebDocumentViewer_AriaLabelPagerCombobox'),
                    'class': 'dx-page-selectbox-popup-wrapper'
                }
            })
                .generateProperty('inputAttr', {
                'aria-label': getLocalization('Pager', 'ASPxReportsStringId.WebDocumentViewer_AriaLabelPagerCombobox')
            })
                .getViewModel())
                .getViewModel();
            this.subscribe(action.pageItems, (pageItems) => selectBoxViewModel.widget.dataSource = pageItems, action.setPageItemsChangedEvent);
            this.subscribe(action.selectedItem, (selectedItem) => selectBoxViewModel.widget.value = selectedItem, action.setSelectedItemChangedEvent);
            return selectBoxViewModel;
        }
        else if (this._isExportAction(action)) {
            const getItems = (items) => {
                return items.map(x => ({
                    imageClassName: x.imageClassName,
                    imageTemplateName: x.imageTemplateName,
                    text: getLocalization(x.text, x.textId),
                    format: x.format,
                    items: x.items && getItems(x.items) || undefined
                }));
            };
            const exportActionViewModel = createViewModelGenerator(viewModel)
                .generateProperty('widget', createViewModelGenerator()
                .generateProperty('items', getItems(action.items))
                .generateProperty('onItemRendered', (e) => action.eventHandlers.onItemRendered(e))
                .generateProperty('onSubmenuHiding', (e) => action.eventHandlers.onSubmenuHiding(e))
                .generateProperty('onSubmenuShown', (e) => action.eventHandlers.onSubmenuShown(e))
                .generateProperty('onSubmenuShowing', (container, element) => action.eventHandlers.onSubmenuShowing(container, element))
                .getViewModel())
                .generateProperty('getPopupContainer', getParentContainer)
                .getViewModel();
            this.subscribe(action.items, (items) => exportActionViewModel.widget.items = getItems(items), action.setItemsChangedEvent);
            return exportActionViewModel;
        }
        return viewModel;
    }
    onPropertyChanged(args) {
        if (args.propertyName === 'globalActionProviders')
            this.updateToolbarItems();
    }
    processShortcut(actions, e) {
        if (this.shouldIgnoreProcessing(e))
            return;
        super.processShortcut(actions, e);
    }
    dispose() {
        super.dispose();
        this.globalActionProviders = [];
        this.removeProperties();
    }
}
__decorate([
    mutableArray(() => [])
], ActionLists.prototype, "globalActionProviders", void 0);
export class PreviewActions extends Disposable {
    constructor(reportPreview) {
        super();
        this.actions = [];
        const exportDisabledChangedCallbacks = [];
        const visibilityChangedCallbacks = [];
        const zoomChangedCallbacks = [];
        const editingFieldsHighlightedChangedCallbacks = [];
        const editingFieldsChangedCallbacks = [];
        const showMultipagePreviewChangedCallbacks = [];
        const pageIndexChangedCallbacks = [];
        const pagesChangedCallbacks = [];
        const pagesAndIndexChangedCallbacks = [];
        this.addDisposable(reportPreview.events.on('zoomChanged', (args) => {
            zoomChangedCallbacks.forEach(event => event(args.newValue));
        }), reportPreview.events.on('editingFieldsHighlightedChanged', (args) => {
            editingFieldsHighlightedChangedCallbacks.forEach(event => event(args.newValue));
        }), reportPreview.events.on('showMultipagePreviewChanged', (args) => {
            showMultipagePreviewChangedCallbacks.forEach(event => event(args.newValue));
        }), reportPreview.events.on('pageIndexChanged', (args) => {
            pageIndexChangedCallbacks.forEach(event => event(args.newValue));
            pagesAndIndexChangedCallbacks.forEach(event => event(reportPreview.pages, reportPreview.pageIndex));
        }), reportPreview.events.on('pagesChanged', (args) => {
            pagesChangedCallbacks.forEach(event => event(reportPreview.pages));
            pagesAndIndexChangedCallbacks.forEach(event => event(reportPreview.pages, reportPreview.pageIndex));
        }), reportPreview.events.on('previewVisibleChanged', (args) => {
            visibilityChangedCallbacks.forEach(event => event(args.newValue));
        }), reportPreview.events.on('exportDisabledChanged', (args) => {
            exportDisabledChangedCallbacks.forEach(event => event(args.newValue));
        }), reportPreview.events.on('_editingFieldsChanged', (args) => {
            editingFieldsChangedCallbacks.forEach(event => event(args.newValue));
        }));
        const goBackwardsDisabled = () => reportPreview.pageIndex < 1;
        const setVisibleChangedEvent = (callback) => {
            const handler = (visible) => callback(visible);
            visibilityChangedCallbacks.push(handler);
            return () => visibilityChangedCallbacks.splice(visibilityChangedCallbacks.indexOf(handler), 1);
        };
        this.actions.push({
            id: ActionId.FirstPage,
            text: getLocalization('First Page', 'ASPxReportsStringId.DocumentViewer_RibbonCommandText_FirstPage'),
            imageClassName: 'dxrd-image-preview-first',
            imageTemplateName: 'dxrd-svg-preview-first_page',
            setDisabledChangedEvent: (callback) => {
                const handler = () => callback(goBackwardsDisabled());
                pageIndexChangedCallbacks.push(handler);
                return () => pageIndexChangedCallbacks.splice(pageIndexChangedCallbacks.indexOf(handler), 1);
            },
            getDisabled: () => goBackwardsDisabled(),
            getVisible: () => reportPreview.previewVisible,
            setVisibleChangedEvent: setVisibleChangedEvent,
            hotKey: { ctrlKey: true, keyCode: 37 },
            clickAction: () => {
                if (reportPreview.pageIndex > 0) {
                    reportPreview.goToPage(0);
                }
            }
        });
        this.actions.push({
            id: ActionId.PrevPage,
            text: getLocalization('Previous Page', 'ASPxReportsStringId.ToolBarItemText_PreviousPage'),
            imageClassName: 'dxrd-image-preview-prev',
            imageTemplateName: 'dxrd-svg-preview-previous_page',
            setDisabledChangedEvent: (callback) => {
                const handler = () => callback(goBackwardsDisabled());
                pageIndexChangedCallbacks.push(handler);
                return () => pageIndexChangedCallbacks.splice(pageIndexChangedCallbacks.indexOf(handler), 1);
            },
            getDisabled: () => goBackwardsDisabled(),
            getVisible: () => reportPreview.previewVisible,
            setVisibleChangedEvent: setVisibleChangedEvent,
            hotKey: { ctrlKey: false, keyCode: 37 },
            clickAction: () => {
                if (reportPreview.pageIndex >= 1) {
                    reportPreview.goToPage(reportPreview.pageIndex - 1, false, 500);
                }
            },
        });
        const getPageItems = () => {
            var pageCount = reportPreview.pages.length;
            if (pageCount === 0 || reportPreview.pageIndex === -1) {
                return [];
            }
            var pagesArray = new Array();
            for (var i = 0; i < pageCount;) {
                pagesArray.push({ index: i, text: ++i });
            }
            return {
                store: pagesArray,
                paginate: pageCount > 200,
                pageSize: 100
            };
        };
        const getSelectedItem = () => {
            if (reportPreview.pageIndex < 0) {
                return null;
            }
            var items = getPageItems();
            return items && items.store && (items.store.length > reportPreview.pageIndex) && items.store[reportPreview.pageIndex];
        };
        const previewHasNoPages = () => reportPreview.pages.length === 0 || reportPreview.pageIndex === -1;
        var paginationSelectBoxViewModel = {
            id: ActionId.Pagination,
            text: 'Pagination',
            setDisabledChangedEvent: (callback) => {
                const handler = () => callback(previewHasNoPages());
                pagesAndIndexChangedCallbacks.push(handler);
                return () => pagesAndIndexChangedCallbacks.splice(pagesAndIndexChangedCallbacks.indexOf(handler), 1);
            },
            getDisabled: () => previewHasNoPages(),
            getVisible: () => reportPreview.previewVisible,
            setVisibleChangedEvent: setVisibleChangedEvent,
            selectItem: (newValue) => {
                if (!!newValue && (newValue.index || newValue.index === 0)) {
                    reportPreview.goToPage(newValue.index);
                }
            },
            selectedItem: getSelectedItem(),
            setSelectedItemChangedEvent: (callback) => {
                const handler = () => callback(getSelectedItem());
                pagesAndIndexChangedCallbacks.push(handler);
                return () => pagesAndIndexChangedCallbacks.splice(pagesAndIndexChangedCallbacks.indexOf(handler), 1);
            },
            pageItems: getPageItems(),
            setPageItemsChangedEvent: (callback) => {
                const handler = () => callback(getPageItems());
                pagesChangedCallbacks.push(handler);
                return () => pagesChangedCallbacks.splice(pagesChangedCallbacks.indexOf(handler), 1);
            },
            _isPageChanged: (value) => {
                var val = parseInt && parseInt(value);
                if (!!val && val-- > 0 && val < reportPreview.pages.length) {
                    reportPreview.goToPage(val);
                    return true;
                }
                return false;
            },
            displayExpr: function (value) {
                if (reportPreview.pageIndex !== -1 && value && reportPreview.pageIndex !== value.index) {
                    return value.text;
                }
                if (reportPreview.pageIndex === -1) {
                    return getLocalization('0 pages', 'ASPxReportsStringId.WebDocumentViewer_0Pages');
                }
                else {
                    var ofText = getLocalization('of', 'ASPxReportsStringId.ToolBarItemText_OfLabel');
                    return (reportPreview.pageIndex + 1) + ' ' + ofText + ' ' + reportPreview.pages.length;
                }
            },
            itemTemplate: function (value) { return value.text; },
            searchMode: 'startswith',
            searchTimeout: 10,
            templateName: 'dxrd-preview-pager'
        };
        const goForwardActionDisabled = () => reportPreview.pageIndex < 0 || reportPreview.pageIndex >= reportPreview.pages.length - 1;
        this.actions.push(paginationSelectBoxViewModel);
        this.actions.push({
            id: ActionId.NextPage,
            text: getLocalization('Next Page', 'ASPxReportsStringId.ToolBarItemText_NextPage'),
            imageClassName: 'dxrd-image-preview-next',
            imageTemplateName: 'dxrd-svg-preview-next_page',
            setDisabledChangedEvent: (callback) => {
                const handler = () => callback(goForwardActionDisabled());
                pagesAndIndexChangedCallbacks.push(handler);
                return () => pagesAndIndexChangedCallbacks.splice(pagesAndIndexChangedCallbacks.indexOf(handler), 1);
            },
            getDisabled: () => goForwardActionDisabled(),
            getVisible: () => reportPreview.previewVisible,
            setVisibleChangedEvent: setVisibleChangedEvent,
            hotKey: { ctrlKey: false, keyCode: 39 },
            clickAction: () => {
                if (reportPreview.pageIndex < reportPreview.pages.length - 1) {
                    reportPreview.goToPage(reportPreview.pageIndex + 1, false, 500);
                }
            }
        });
        this.actions.push({
            id: ActionId.LastPage,
            text: getLocalization('Last Page', 'ASPxReportsStringId.DocumentViewer_RibbonCommandText_LastPage'),
            imageClassName: 'dxrd-image-preview-last',
            imageTemplateName: 'dxrd-svg-preview-last_page',
            setDisabledChangedEvent: (callback) => {
                const handler = () => callback(goForwardActionDisabled());
                pagesAndIndexChangedCallbacks.push(handler);
                return () => pagesAndIndexChangedCallbacks.splice(pagesAndIndexChangedCallbacks.indexOf(handler), 1);
            },
            getDisabled: () => goForwardActionDisabled(),
            getVisible: () => reportPreview.previewVisible,
            setVisibleChangedEvent: setVisibleChangedEvent,
            hotKey: { ctrlKey: true, keyCode: 39 },
            clickAction: () => {
                if (reportPreview.pageIndex < reportPreview.pages.length - 1) {
                    reportPreview.goToPage(reportPreview.pages.length - 1);
                }
            }
        });
        const getMultipageToggleImageClassName = (showMultipagePreview) => {
            return showMultipagePreview ? 'dxrd-image-preview-single-page' : 'dxrd-image-preview-multipage';
        };
        const getMultipageToggleImageTemplateName = (showMultipagePreview) => {
            return showMultipagePreview ? 'dxrd-svg-preview-single_page' : 'dxrd-svg-preview-multi_page_preview';
        };
        const multipageModeAction = {
            id: ActionId.MultipageToggle,
            text: getLocalization('Toggle Multipage Mode', 'ASPxReportsStringId.WebDocumentViewer_ToggleMultipageMode'),
            imageClassName: getMultipageToggleImageClassName(reportPreview.showMultipagePreview),
            imageTemplateName: getMultipageToggleImageTemplateName(reportPreview.showMultipagePreview),
            getVisible: () => reportPreview.previewVisible,
            setVisibleChangedEvent: setVisibleChangedEvent,
            hotKey: { ctrlKey: true, keyCode: 77 },
            clickAction: () => {
                var zoom = reportPreview.originalZoom;
                reportPreview.showMultipagePreview = !reportPreview.showMultipagePreview;
                reportPreview.zoom = zoom;
            },
            setImageClassNameChangedEvent: (callback) => {
                const handler = (newValue) => callback(getMultipageToggleImageClassName(newValue));
                showMultipagePreviewChangedCallbacks.push(handler);
                return () => showMultipagePreviewChangedCallbacks.splice(showMultipagePreviewChangedCallbacks.indexOf(handler), 1);
            },
            setImageTemplateNameChangedEvent: (callback) => {
                const handler = (newValue) => callback(getMultipageToggleImageTemplateName(newValue));
                showMultipagePreviewChangedCallbacks.push(handler);
                return () => showMultipagePreviewChangedCallbacks.splice(showMultipagePreviewChangedCallbacks.indexOf(handler), 1);
            },
            hasSeparator: true
        };
        this.actions.push(multipageModeAction);
        this.actions.push({
            id: ActionId.ZoomOut,
            text: getLocalization('Zoom Out', 'DevExpress.XtraPrinting.PrintingSystemCommand.ZoomOut'),
            imageClassName: 'dxrd-image-zoomout',
            imageTemplateName: 'dxrd-svg-toolbar-zoomout',
            visible: true,
            hotKey: { ctrlKey: false, keyCode: 109 },
            clickAction: () => {
                var currentZoom = reportPreview.zoom;
                var zoomLevel = currentZoom > 0 ? currentZoom : reportPreview.originalZoom;
                reportPreview.zoom = Math.max(zoomLevel - reportPreview.zoomStep, 0.1);
            },
            hasSeparator: true
        });
        const getZoomLevels = (showMultipagePreview) => {
            var items = reportPreview.predefinedZoomLevels.slice(0);
            if (showMultipagePreview === false && items.indexOf(0) === -1) {
                items.push(ZoomAutoBy.PageWidth);
                items.push(ZoomAutoBy.WholePage);
            }
            return items;
        };
        const zoomAction = {
            id: ActionId.ZoomSelector,
            visible: true,
            hotKey: { ctrlKey: true, keyCode: 187 },
            templateName: 'dxrd-zoom-autofit-select-template',
            displayExpr: (val) => {
                if (val === ZoomAutoBy.PageWidth) {
                    return getLocalization('Page Width', 'DevExpress.XtraReports.UI.XtraReport.PageWidth');
                }
                else if (Math.round(val * 100) === 0) {
                    return getLocalization('Whole Page', 'PreviewStringId.MenuItem_ZoomWholePage');
                }
                else {
                    return Math.round((val || reportPreview.zoom) * 100) + '%';
                }
            },
            onCustomItemCreating: (e) => { e.customItem = parseZoom(e.text); },
            zoom: reportPreview.zoom,
            zoomLevels: getZoomLevels(reportPreview.showMultipagePreview),
            setZoomLevelsChangedEvent: (callback) => {
                const handler = (newValue) => callback(getZoomLevels(newValue));
                showMultipagePreviewChangedCallbacks.push(handler);
                return () => showMultipagePreviewChangedCallbacks.splice(showMultipagePreviewChangedCallbacks.indexOf(handler), 1);
            },
            setZoom: (newValue) => {
                reportPreview.zoom = newValue;
            },
            setZoomChangedEvent: (callback) => {
                zoomChangedCallbacks.push(callback);
                return () => zoomChangedCallbacks.splice(zoomChangedCallbacks.indexOf(callback), 1);
            }
        };
        this.actions.push(zoomAction);
        this.actions.push({
            id: ActionId.ZoomIn,
            text: getLocalization('Zoom In', 'DevExpress.XtraPrinting.PrintingSystemCommand.ZoomIn'),
            imageClassName: 'dxrd-image-zoomin',
            imageTemplateName: 'dxrd-svg-toolbar-zoomin',
            visible: true,
            hotKey: { ctrlKey: false, keyCode: 107 },
            clickAction: () => {
                var currentZoom = reportPreview.zoom;
                var zoomLevel = currentZoom > 0 ? currentZoom : reportPreview.originalZoom;
                reportPreview.zoom = Math.min(zoomLevel + reportPreview.zoomStep, 10);
            }
        });
        const getHighlightEditingFieldsVisible = () => EditablePreviewEnabled() && reportPreview.previewVisible;
        const getHighlightEditingFieldsDisabled = () => reportPreview.editingFieldsProvider().length < 1;
        const highlightingAction = {
            id: ActionId.HighlightEditingFields,
            text: getLocalization('Highlight Editing Fields', 'DevExpress.XtraPrinting.PrintingSystemCommand.HighlightEditingFields'),
            imageClassName: 'dxrp-image-hightlight-editing-fields',
            imageTemplateName: 'dxrd-svg-toolbar-hightlightEditingFields',
            setDisabledChangedEvent: (callback) => {
                const handler = () => callback(getHighlightEditingFieldsDisabled());
                editingFieldsChangedCallbacks.push(handler);
                return () => editingFieldsChangedCallbacks.splice(editingFieldsChangedCallbacks.indexOf(handler), 1);
            },
            getDisabled: () => getHighlightEditingFieldsDisabled(),
            getVisible: () => getHighlightEditingFieldsVisible(),
            setVisibleChangedEvent: (callback) => {
                const handler = () => callback(getHighlightEditingFieldsVisible());
                visibilityChangedCallbacks.push(handler);
                const subscriptionDispose = EditablePreviewEnabled.subscribe(handler);
                return () => {
                    subscriptionDispose();
                    visibilityChangedCallbacks.splice(visibilityChangedCallbacks.indexOf(handler), 1);
                };
            },
            selected: reportPreview.editingFieldsHighlighted,
            setSelectedChangedEvent: (callback) => {
                editingFieldsHighlightedChangedCallbacks.push(callback);
                return () => editingFieldsHighlightedChangedCallbacks.splice(editingFieldsHighlightedChangedCallbacks.indexOf(callback), 1);
            },
            hotKey: { ctrlKey: true, keyCode: 72 },
            clickAction: () => {
                reportPreview.editingFieldsHighlighted = !reportPreview.editingFieldsHighlighted;
            },
            hasSeparator: true
        };
        this.actions.push(highlightingAction);
        this.addDisposable(reportPreview.events.on('propertyChanged', (e) => {
            if (e.propertyName === 'showMultipagePreview') {
                zoomAction.zoomLevels = getZoomLevels(reportPreview.showMultipagePreview);
                multipageModeAction.imageClassName = getMultipageToggleImageClassName(reportPreview.showMultipagePreview);
                multipageModeAction.imageTemplateName = getMultipageToggleImageTemplateName(reportPreview.showMultipagePreview);
            }
            else if (e.propertyName === 'zoom') {
                zoomAction.zoom = reportPreview.zoom;
            }
            else if (e.propertyName === 'editingFieldsHighlighted') {
                highlightingAction.selected = reportPreview.editingFieldsHighlighted;
            }
        }));
        this.actions.push({
            id: ActionId.Print,
            text: getLocalization('Print', 'ASPxReportsStringId.DocumentViewer_RibbonPrintGroupText'),
            imageClassName: 'dxrd-image-print',
            imageTemplateName: 'dxrd-svg-preview-print',
            hasSeparator: true,
            getDisabled: () => reportPreview.exportDisabled,
            setDisabledChangedEvent: (callback) => {
                exportDisabledChangedCallbacks.push(callback);
                return () => exportDisabledChangedCallbacks.splice(exportDisabledChangedCallbacks.indexOf(callback), 1);
            },
            visible: true,
            hotKey: { ctrlKey: true, keyCode: 80 },
            clickAction: () => {
                if (reportPreview.exportDisabled) {
                    return;
                }
                reportPreview.printDocument();
            }
        });
        this.actions.push({
            id: ActionId.PrintPage,
            text: getLocalization('Print Page', 'ASPxReportsStringId.DocumentViewer_RibbonCommandText_PrintPage'),
            imageClassName: 'dxrd-image-print-page',
            imageTemplateName: 'dxrd-svg-preview-print_page',
            getDisabled: () => reportPreview.exportDisabled,
            setDisabledChangedEvent: (callback) => {
                exportDisabledChangedCallbacks.push(callback);
                return () => exportDisabledChangedCallbacks.splice(exportDisabledChangedCallbacks.indexOf(callback), 1);
            },
            visible: true,
            clickAction: () => {
                if (reportPreview.exportDisabled) {
                    return;
                }
                reportPreview.printDocument(reportPreview.pageIndex);
            }
        });
    }
    wrapDisposable(object) {
        this.addDisposable(object);
        return object;
    }
    dispose() {
        super.dispose();
        this.removeProperties();
    }
    getActions(context) {
        return this.actions;
    }
}
