// @devexpress/analytics-core/core/tools/_actionListBase.js
/**
* DevExpress Analytics (core\tools\_actionListBase.js)
* Version:  23.2.3
* Build date: Dec 1, 2023
* Copyright (c) 2012 - 2023 Developer Express Inc. ALL RIGHTS RESERVED
* License: https://www.devexpress.com/Support/EULAs/universal.xml
*/
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
import * as $ from 'jquery';
import { BaseRenderingModel, mutable } from '../../serializer/native/models/base.model';
import { koUtils } from '../utils/_koUtils';
import { createActionListBaseViewModel, createBaseActionViewModel } from './_actionListBase.viewModel';
export class ActionListsBase extends BaseRenderingModel {
    constructor(enabled) {
        super();
        this.toolbarItems = [];
        this.enabled = enabled || (() => true);
    }
    createViewModel() {
        return createActionListBaseViewModel.call(this, super.createViewModel());
    }
    subscribeOnChanges(action, viewModelAction, propertyNames) {
        propertyNames.forEach((propertyName) => {
            if (koUtils.isSubscribable(action[propertyName]))
                this.subscribe(action[propertyName], () => {
                    viewModelAction[propertyName] = koUtils.unwrap(action[propertyName]);
                });
        });
    }
    subscribe(property, callback, valueChangedEventSetter) {
        if (koUtils.isSubscribable(property))
            this._disposables.push(property.subscribe((newValue) => {
                callback(newValue);
                this.events.call('propertyChanged');
            }));
        else if (valueChangedEventSetter) {
            this._disposables.push({
                dispose: valueChangedEventSetter((newValue) => callback(newValue))
            });
        }
    }
    _isActionDisabled(action) {
        return action.getDisabled ? action.getDisabled() : koUtils.unwrap(action.disabled);
    }
    _isActionVisible(action) {
        return action.getVisible ? action.getVisible() : koUtils.unwrap(action.visible);
    }
    createActionViewModel(action, index) {
        return createBaseActionViewModel.call(this, action, index);
    }
    onPropertyChanged(args) {
    }
    updateViewModel(args) {
        if (args.propertyName === 'toolbarItems')
            this.getViewModel().toolbarItems = this.mapActionToViewModels(koUtils.unwrap(this.toolbarItems));
    }
    mapActionToViewModels(actions) {
        return actions.map((x, i) => this.createActionViewModel(x, i));
    }
    processShortcut(actions, e) {
        for (let i = 0; i < actions.length; i++) {
            const action = actions[i];
            if (action.hotKey && action.hotKey.ctrlKey === e.ctrlKey && action.hotKey.keyCode === e.keyCode) {
                const enabled = !this._isActionDisabled(action);
                const visible = this._isActionVisible(action);
                if (enabled && visible) {
                    action.clickAction();
                    e.preventDefault();
                }
            }
        }
    }
    shouldIgnoreProcessing(e) {
        if (e.altKey || !this.enabled())
            return true;
        const activeElement = $.fn.constructor(document.activeElement);
        if (activeElement.is('textarea') || activeElement.is(':input') && (['password', 'text', 'number'].indexOf(activeElement.attr('type')) != -1)) {
            return true;
        }
        return false;
    }
}
__decorate([
    mutable(() => [])
], ActionListsBase.prototype, "toolbarItems", void 0);
