// devexpress-reporting/designer/bands/metadata/xrBandMetaData.js
/**
* DevExpress HTML/JS Reporting (designer\bands\metadata\xrBandMetaData.js)
* Version:  23.2.3
* Build date: Dec 1, 2023
* Copyright (c) 2012 - 2023 Developer Express Inc. ALL RIGHTS RESERVED
* License: https://www.devexpress.com/Support/EULAs/universal.xml
*/
import { parseBool } from '@devexpress/analytics-core/analytics-utils';
import { keepTogetherDefaultValueFalse, textAlignment } from '../../controls/metadata/properties/metadata';
import { commonBandProperties, fontGroup } from '../../controls/metadata/properties/metadataGroups';
import { commonBandScripts } from '../../controls/metadata/properties/scriptMetadata';
import { styleName, stylePriority, stylesObj } from '../../controls/metadata/properties/style';
import { height, pageBreak, printAcrossBands, printAtBottom } from './bandsMetadata';
export var expanded = { propertyName: 'expanded', modelName: '@Expanded', from: parseBool, defaultVal: true };
export var commonBandSerializationInfo = [
    textAlignment, expanded,
    { propertyName: 'controls', modelName: 'Controls', array: true },
    { propertyName: 'bands', modelName: 'SubBands', array: true },
].concat(commonBandProperties, fontGroup);
export var bandSerializationInfo = [
    styleName, stylesObj, stylePriority, height, printAcrossBands
].concat(commonBandSerializationInfo);
export var reportHeaderBandSerializationInfo = [keepTogetherDefaultValueFalse, pageBreak, commonBandScripts].concat(bandSerializationInfo);
export var reportFooterBandSerializationInfo = [printAtBottom].concat(reportHeaderBandSerializationInfo);
export var popularPropertiesReportHeader = [pageBreak.propertyName, 'keepTogether'];
export var popularPropertiesReportFooter = [pageBreak.propertyName, 'keepTogether', 'printAtBottom'];
