// devexpress-reporting/designer/helpers/reportParameterHelper.js
/**
* DevExpress HTML/JS Reporting (designer\helpers\reportParameterHelper.js)
* Version:  23.2.3
* Build date: Dec 1, 2023
* Copyright (c) 2012 - 2023 Developer Express Inc. ALL RIGHTS RESERVED
* License: https://www.devexpress.com/Support/EULAs/universal.xml
*/
import { getUniqueNameForNamedObjectsArray } from '@devexpress/analytics-core/analytics-internal';
import * as ko from 'knockout';
import { ParameterHelper } from '../../viewer/parameters/parameterHelper';
import { GroupLayoutItem, ParameterLayoutItem } from '../dataObjects/parameters/layoutItems';
export class ReportParameterHelper extends ParameterHelper {
    constructor(container) {
        super();
        this.container = container;
        this._disposables.push(this.allLayoutItems = ko.pureComputed(() => {
            var _a;
            var allItems = [];
            if (this.parameterPanelLayoutItems && ((_a = this.parameterPanelLayoutItems()) === null || _a === void 0 ? void 0 : _a.length))
                allItems = this.getAllLayoutItems(this.parameterPanelLayoutItems());
            return allItems;
        }));
    }
    get parameters() {
        return this.container.parameters;
    }
    get parameterPanelLayoutItems() {
        return this.container.parameterPanelLayoutItems;
    }
    getAllLayoutItems(items) {
        var allChildLayoutItems = [];
        items.forEach((item) => {
            allChildLayoutItems.push(item);
            if (item instanceof GroupLayoutItem) {
                var subLayoutItems = this.getAllLayoutItems(item.parameterPanelLayoutItems());
                if (subLayoutItems && subLayoutItems.length)
                    allChildLayoutItems.push(...subLayoutItems);
            }
        });
        return allChildLayoutItems;
    }
    addParameterPanelLayoutItem(item) {
        if (item instanceof GroupLayoutItem && !item.title()) {
            var newName = getUniqueNameForNamedObjectsArray(this.allLayoutItems().filter(x => x instanceof GroupLayoutItem), 'Group');
            item.title(newName);
        }
    }
    startEditing() {
        this.parameters().forEach(parameter => {
            parameter._isEditing(true);
            parameter._showLayoutProperties(true);
        });
    }
    endEditing() {
        this.parameters().forEach(parameter => {
            parameter._isEditing(false);
            parameter._showLayoutProperties(false);
        });
    }
    updateParameterLayoutItems() {
        this.parameters().forEach(parameter => {
            var parameterLayoutItem = this.getParameterLayoutItem(parameter);
            if (!parameterLayoutItem) {
                var newItem = new ParameterLayoutItem({}, this.container, undefined, parameter);
                this.parameterPanelLayoutItems.push(newItem);
            }
        });
    }
    clearLayoutItems() {
        var toRemove = [];
        this.parameterPanelLayoutItems().forEach(item => {
            if (item instanceof ParameterLayoutItem) {
                var isModify = item.getInfo().some(x => !!x.editor && item.isPropertyModified(x.propertyName));
                if (!isModify) {
                    toRemove.push(item);
                }
            }
        });
        toRemove.forEach(x => this.parameterPanelLayoutItems.remove(x));
    }
    getParameterLayoutItem(parameter) {
        return this.allLayoutItems().filter(x => x['parameter'] && x['parameter']() === parameter)[0];
    }
    removeParameterModel(parameter) {
        for (var parameterLayoutItem of this.allLayoutItems()) {
            if (parameterLayoutItem['parameter'] && parameterLayoutItem['parameter']() === parameter) {
                parameterLayoutItem.delete();
            }
        }
    }
}
