// @devexpress/analytics-core/serializer/native/viewModels/viewModelGenerator.js
/**
* DevExpress Analytics (serializer\native\viewModels\viewModelGenerator.js)
* Version:  23.2.3
* Build date: Dec 1, 2023
* Copyright (c) 2012 - 2023 Developer Express Inc. ALL RIGHTS RESERVED
* License: https://www.devexpress.com/Support/EULAs/universal.xml
*/
export let createViewModelGenerator = function (_viewModel = {}) {
    const generator = {
        createDefaultModel: (model) => {
            _viewModel = {
                getModel() {
                    return model;
                }
            };
            return generator;
        },
        generateProperty(propertyName, value, setter) {
            let model = _viewModel;
            model[propertyName] = value;
            if (setter) {
                model['set_' + propertyName] = (newValue) => setter(newValue);
            }
            return this;
        },
        configureProperty(propertyName, configure) {
            if (propertyName in _viewModel)
                configure(_viewModel[propertyName]);
            return this;
        },
        getViewModel() {
            return _viewModel;
        }
    };
    return generator;
};
export function setDefaultViewModelGenerator(modelGenerator) {
    createViewModelGenerator = modelGenerator;
}
