// devexpress-reporting/designer/services/_reportScriptService.js
/**
* DevExpress HTML/JS Reporting (designer\services\_reportScriptService.js)
* Version:  23.2.3
* Build date: Dec 1, 2023
* Copyright (c) 2012 - 2023 Developer Express Inc. ALL RIGHTS RESERVED
* License: https://www.devexpress.com/Support/EULAs/universal.xml
*/
import { ajax } from '@devexpress/analytics-core/analytics-internal';
import * as $ from 'jquery';
import { HandlerUri } from '../utils/settings';
export class ReportScriptService {
    static validateScripts(report) {
        return ajax(HandlerUri(), 'validateScripts', encodeURIComponent(JSON.stringify({
            report: JSON.stringify({
                'XtraReportsLayoutSerializer': report.serialize()
            })
        })));
    }
    static getCompletions(editor, session, pos, prefix, callback, report, editorInstance, guid) {
        if (guid) {
            return ajax(HandlerUri(), 'getCompletions', encodeURIComponent(JSON.stringify({
                Line: pos.row,
                Column: pos.column,
                Guid: guid,
                Script: editorInstance.getValue()
            })));
        }
        else {
            return $.Deferred().resolve().promise();
        }
    }
    static setCodeDom(key, reportLayout) {
        return ajax(HandlerUri(), 'setReportLayout', encodeURIComponent(JSON.stringify({
            ReportLayout: reportLayout,
            Key: key
        })));
    }
}
