// devexpress-reporting/designer/bands/xrSubband.js
/**
* DevExpress HTML/JS Reporting (designer\bands\xrSubband.js)
* Version:  23.2.3
* Build date: Dec 1, 2023
* Copyright (c) 2012 - 2023 Developer Express Inc. ALL RIGHTS RESERVED
* License: https://www.devexpress.com/Support/EULAs/universal.xml
*/
import * as ko from 'knockout';
import { pageBreak, printAcrossBands } from './metadata/bandsMetadata';
import { BandSurface, BandViewModel } from './xrBand';
export class SubBandViewModel extends BandViewModel {
    constructor(band, parent, serializer) {
        super(band, parent, serializer);
    }
    isPropertyDisabled(name) {
        if (name === pageBreak.propertyName)
            return this.controlType === 'SubBand' && this[printAcrossBands.propertyName] && this[printAcrossBands.propertyName]();
        else
            return super.isPropertyDisabled(name);
    }
    isAllowedParent(target) {
        return target instanceof BandViewModel;
    }
}
export class SubBandSurface extends BandSurface {
    constructor() {
        super(...arguments);
        this.leftMarginTemplate = 'dxrd-sub-band-coordinate-grid';
    }
    getAbsolutePositionY() {
        var y = super.getAbsolutePositionY();
        if (this.parent.bandsHolder.bands().indexOf(this) === 0) {
            return y + (this.parent.heightWithoutSubBands());
        }
        return y;
    }
    getBackgroundRect() {
        var top = 0, bottom, height = this._height();
        var parent = this.parent;
        var parentBands = ko.unwrap(parent.bandsHolder.bands);
        var parentBackgroundRect = ko.unwrap(parent.backgroundRect);
        top += (parentBackgroundRect.top + parentBackgroundRect.height);
        bottom = parentBackgroundRect.bottom;
        var bandIndex = parentBands.indexOf(this);
        for (var i = 0; i < bandIndex; i++) {
            top += parentBands[i]._totalHeight();
        }
        if (top > bottom)
            height = 0;
        else if (top + height > bottom)
            height = bottom - top;
        return { top, bottom, height };
    }
    _initMultiColumn() {
        this._disposables.push(this.multiColumn = ko.computed(() => {
            if (this.parent.multiColumn && this.parent.multiColumn() && this.parent.multiColumn().haveColumns()) {
                return this.parent.multiColumn();
            }
        }));
    }
    get parent() {
        return this._getParent();
    }
}
