// devexpress-reporting/designer/dataObjects/objectStorage.js
/**
* DevExpress HTML/JS Reporting (designer\dataObjects\objectStorage.js)
* Version:  23.2.3
* Build date: Dec 1, 2023
* Copyright (c) 2012 - 2023 Developer Express Inc. ALL RIGHTS RESERVED
* License: https://www.devexpress.com/Support/EULAs/universal.xml
*/
import { Disposable } from '@devexpress/analytics-core/analytics-utils';
import { ObjectStorageItem } from './objectStorageItem';
import { DynamicListLookUpSettings, StaticListLookUpSettings } from './parameters/lookupSettings';
import { RangeEndParameter, RangeParametersSettings, RangeStartParameter } from './parameters/rangeSettings';
export class ObjectsStorage extends Disposable {
    constructor(objects, dsHelperProvider) {
        super();
        this.objects = objects;
        this.dsHelperProvider = dsHelperProvider;
    }
    findType(content) {
        var objectStorages = this.objects().filter((item) => { return item instanceof ObjectStorageItem; });
        var result = objectStorages.filter((item) => {
            return !!item.type && item.type() === 'System.Type' && item.content() === content;
        });
        return result.length === 0 ? null : result[0];
    }
    getType(type) {
        var typeObject = this.findType(type);
        if (!typeObject) {
            typeObject = new ObjectStorageItem({
                '@Content': type,
                '@Type': 'System.Type'
            }, this.dsHelperProvider);
            this.objects.push(typeObject);
        }
        return typeObject;
    }
    addValue() {
        var newValueRef = new ObjectStorageItem({ '@Content': '' }, this.dsHelperProvider);
        this.objects.push(newValueRef);
        return newValueRef;
    }
    createStaticLookUpSetting() {
        var lookUpObject = new StaticListLookUpSettings({}, this.dsHelperProvider);
        this.objects.push(lookUpObject);
        return lookUpObject;
    }
    createDynamicLookUpSetting() {
        var lookUpObject = new DynamicListLookUpSettings({}, this.dsHelperProvider);
        this.objects.push(lookUpObject);
        return lookUpObject;
    }
    createRangeSetting() {
        var rangeSettingsObject = new RangeParametersSettings({}, this.dsHelperProvider);
        this.objects.push(rangeSettingsObject);
        var startParameter = new RangeStartParameter({}, this.dsHelperProvider);
        rangeSettingsObject.startParameter(startParameter);
        this.objects.push(startParameter);
        var endParameter = new RangeEndParameter({}, this.dsHelperProvider);
        rangeSettingsObject.endParameter(endParameter);
        this.objects.push(endParameter);
        return rangeSettingsObject;
    }
}
